src/psyclone/dynamo0p3.py

```diff
@@ -6514,62 +6514,61 @@ def _halo_read_access(self, arg):
-                # compute annexed dofs and True if we don't (as
-                # annexed dofs are part of the level 1 halo).
-                return not Config.get()\
-                                 .api_conf("dynamo0.3").compute_annexed_dofs
-            if self._upper_bound_name in ["ndofs"]:
-                # Argument does not read from the halo
+        if arg.is_field:
+            # This is a field so might read from a halo
+            if arg.access in [AccessType.WRITE]:
+                # This is not a read access
+                return False
+            if arg.access in AccessType.all_read_accesses():
+                # This is a read access
+                if arg.descriptor.stencil:
+                    if self._upper_bound_name not in ["cell_halo", "ncells"]:
+                        raise GenerationError(
+                            "Loop bounds other than 'cell_halo' and 'ncells' "
+                            "are currently unsupported for kernels with "
+                            "stencil accesses. Found '{0}'."
+                            "".format(self._upper_bound_name))
+                    # An upper bound of 'cell_halo' means that the
+                    # halo might be accessed irrespective of the
+                    # stencil and a stencil read access with upper
+                    # bound 'ncells' might read from the
+                    # halo due to the stencil.
+                    return True
+                # This is a non-stencil read access
+                if self._upper_bound_name in HALO_ACCESS_LOOP_BOUNDS:
+                    # An upper bound that is part of the halo means
+                    # that the halo might be accessed.
+                    return True
+                if not arg.discontinuous and \
+                   self._upper_bound_name in ["ncells", "nannexed"]:
+                    # Annexed dofs may be accessed. Return False if we
+                    # always compute annexed dofs and True if we don't
+                    # (as annexed dofs are part of the level 1 halo).
+                    return not Config.get().api_conf("dynamo0.3").\
+                        compute_annexed_dofs
+                # The halo is not accessed.
                 return False
-            # Nothing should get to here so raise an exception
             raise InternalError(
-                "DynLoop._halo_read_access(): It should not be possible to "
-                "get to here. Loop upper bound name is '{0}' and arg '{1}' "
-                "access is '{2}'.".format(
-                    self._upper_bound_name, arg.name,
-                    arg.access.api_specific_name()))
-
-        # Access is neither a read nor an inc so does not need halo
-        return False
+                "Unexpected field access type '{0}' found.".format(arg.access))
+        raise InternalError(
+            "Expecting arg '{0}' to be an operator, scalar or field."
+            "".format(arg.name))
```

Comment:
Please include what it is in the error message.

Reply:
I didn't do this originally as I thought it would be hard to find this out. Turns out it is easy :-).